src/Microsoft.Health.Fhir.Core/Features/Conformance/SystemConformanceProvider.cs

```diff
@@ -0,0 +1,109 @@
+
+        public SystemConformanceProvider(
+            IModelInfoProvider modelInfoProvider,
+            ISearchParameterDefinitionManager searchParameterDefinitionManager,
+            Func<IScoped<IEnumerable<IProvideCapability>>> capabilityProviders)
+        {
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
+            EnsureArg.IsNotNull(searchParameterDefinitionManager, nameof(searchParameterDefinitionManager));
+            EnsureArg.IsNotNull(capabilityProviders, nameof(capabilityProviders));
+
+            _modelInfoProvider = modelInfoProvider;
+            _searchParameterDefinitionManager = searchParameterDefinitionManager;
+            _capabilityProviders = capabilityProviders;
+        }
+
+        public override async Task<ResourceElement> GetCapabilityStatementAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (_listedCapabilityStatement == null)
+            {
+                await _sem.WaitAsync(cancellationToken);
+
+                try
+                {
+                    if (_listedCapabilityStatement == null)
+                    {
+                        ICapabilityStatementBuilder builder = CapabilityStatementBuilder.Create(_modelInfoProvider, _searchParameterDefinitionManager)
+                            .Update(x =>
+                            {
+                                x.FhirVersion = _modelInfoProvider.SupportedVersion.ToString();
+                                x.Software = new SoftwareComponent
+                                {
+                                    Name = Resources.ServerName,
+                                    Version = Assembly.GetExecutingAssembly().GetName().Version.ToString(),
+                                };
+                            });
+
+                        using (IScoped<IEnumerable<IProvideCapability>> providerFactory = _capabilityProviders())
+                        {
+                            foreach (IProvideCapability provider in providerFactory.Value)
+                            {
+                                provider.Build(builder);
+                            }
+                        }
+
+                        foreach (Action<ListedCapabilityStatement> postConfiguration in _configurationUpdates)
+                        {
+                            builder.Update(statement => postConfiguration(statement));
+                        }
+
+                        _listedCapabilityStatement = builder.Build().ToResourceElement();
+                    }
+                }
+                finally
+                {
+                    _configurationUpdates.Clear();
+                    _sem.Release();
+                }
+            }
+
+            return _listedCapabilityStatement;
+        }
+
+        public void ConfigureOptionalCapabilities(Action<ListedCapabilityStatement> builderAction)
```

Comment:
Is this guaranteed to be called before GetCapabilityStatementAsync is called?

Reply:
I guess you do have guard against it. I wonder if we should consider making sure the capability statement is built as part of the warm up before the server starts taking requests?

Reply:
Makes sense